### PR TITLE
Fixer for pyproject.toml build-system settings

### PR DESCRIPTION
**Pull request by tiran**
- add missing pyproject.toml
- add or update `[build-system] requires`

Requirements in `update_build_requires` are added to `[build-system] requires`. If a requirement name matches an existing name, then the requirement is replaced.

Requirements in `remove_build_requires` are removed from `[build-system] requires`.

Fixes: #260
Fixes: #274

### CODE REVIEW

File: docs/customization.md [hunk 1]
```diff
@@ -191,6 +191,42 @@ in subdirectories, with the filenames prefixed with the source directory name is
 also supported. The newer format, using subdirectories, is preferred because it
 avoids name collisions between variant source trees.
 
+## `project_override` section
+
+The `project_override` configures the `pyproject.toml` auto-fixer. It can
```
Comment (shubhbapna) [inline docs/customization.md:196]: just a suggestion for an alternate name: `pyproject_override`

File: src/fromager/packagesettings.py [hunk 2]
```diff
@@ -577,6 +602,25 @@ def build_ext_parallel(self) -> bool:
...
+                f"{self.package}: applying project_override: "
+                f"{update_build_requires=}, {remove_build_requires=}"
+            )
+            build_dir = self.build_dir(sdist_root_dir)
+            pyproject.PyprojectFix(
+                self.package,
+                build_dir=build_dir,
+                update_build_requires=update_build_requires,
+                remove_build_requires=remove_build_requires,
+            ).run()
```
[the remark this answers is not in the document]
Comment (tiran) [reply src/fromager/packagesettings.py:620]: I don't understand your remark. Can you elaborate, please?

File: src/fromager/pyproject.py [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+        # add / update requirements
+        for reqstr in self.update_requirements:
+            req = Requirement(reqstr)
+            req_map[canonicalize_name(req.name)] = req
+
+        new_requires = sorted(str(req) for req in req_map.values())
+        if set(new_requires) != set(old_requires):
+            # ignore order of items
+            build_system[BUILD_REQUIRES] = new_requires
+            logger.debug(
```
[the remark this answers is not in the document]
Comment (tiran) [reply src/fromager/pyproject.py:112]: Good idea!

File: src/fromager/sources.py [hunk 4]
```diff
@@ -367,11 +367,31 @@ def default_prepare_source(
 ) -> tuple[pathlib.Path, bool]:
     source_root_dir, is_new = unpack_source(ctx, source_filename)
     if is_new:
-        patch_source(ctx, source_root_dir, req)
-        vendor_rust.vendor_rust(req, source_root_dir)
+        default_prepare_new_source(
```
[the remark this answers is not in the document]
Comment (tiran) [reply src/fromager/sources.py:370]: No, I don't have plans to allow overriding. My idea is to provide the default behavior for "is_new" in a public function, so overrides can mimic the default more easily. I'll rename the function.

File: docs/customization.md (hunk 1 above)
Comment (shubhbapna) [reply docs/customization.md:196]: Ah i see. This name works then

File: src/fromager/packagesettings.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (dhellmann) [reply src/fromager/packagesettings.py:620]: We've been trying to keep the settings and the use of the settings decoupled. This introduces a function to the settings class that manipulates the content of a project. I think I'd just make this a function in the pyproject module, and let something call it from there, passing a settings object so the needed values can be looked up before invoking the fixer.

[the remark this answers is not in the document]
Comment (shubhbapna) [reply src/fromager/packagesettings.py:620]: So the settings function will just return the update_build_requires and remove_build_requires field:
```
def project_override(self) -> None:
        """Apply project_overrides"""
        update_build_requires = self._ps.project_override.update_build_requires
        remove_build_requires = self._ps.project_override.remove_build_requires
        if update_build_requires or remove_build_requires:
            logger.debug(
                f"{self.package}: applying project_override: "
                f"{update_build_requires=}, {remove_build_requires=}"
            )
        return (update_build_requires, remove_build_requires)
```
and then in `default_prepare_new_source`:

```python
    pbi = ctx.package_build_info(req)
    patch_source(ctx, source_root_dir, req)
    update_build_requires, remove_build_requires = pbi.project_override()
    pyproject.PyprojectFix(
                canonicalize_name(req.name),
                build_dir=pbi.build_dir(sdist_root_dir),
                update_build_requires=update_build_requires,
                remove_build_requires=remove_build_requires,
            ).run()
    vendor_rust.vendor_rust(req, source_root_dir)
```

File: src/fromager/pyproject.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (tiran) [reply src/fromager/pyproject.py]: It's now logging with `info` level

File: src/fromager/packagesettings.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (tiran) [reply src/fromager/packagesettings.py]: Makes sense. I have implemented the functions as:

```python
pyproject.apply_project_override(
        ctx=ctx,
        req=req,
        sdist_root_dir=source_root_dir,
    )
```

File: src/fromager/sources.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (tiran) [reply src/fromager/sources.py]: The function is now named `prepare_new_source`.

**Review by shubhbapna — APPROVED** [review]

**Pull request merged**